File: tests/test_reliable_transmission.py
import unittest
from reliable_transmission import send_data, receive_data

class TestReliableTransmission(unittest.TestCase):

    def test_send_data(self):
        peer = "peer1"
        data = "Test data"
        result = send_data(peer, data)
        self.assertEqual(result, 'ok')

    def test_receive_data(self):
        data_handler = lambda data: self.assertEqual(data, "Test data")
        result = receive_data(data_handler)
        self.assertEqual(result, 'ok')

if __name__ == '__main__':
    unittest.main()

File: tests/scala/test_reliable_transmission.scala
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers
import scala.concurrent.Await
import scala.concurrent.duration._

class TestReliableTransmission extends AnyFlatSpec with Matchers {

  "ReliableTransmission" should "send data reliably to a peer" in {
    val result = Await.result(ReliableTransmission.sendData("peer1", "Test data"), 5.seconds)
    result shouldBe ()
  }

  it should "receive data reliably" in {
    val dataHandler: String => Unit = data => data shouldBe "Test data"
    val result = Await.result(ReliableTransmission.receiveData(dataHandler), 5.seconds)
    result shouldBe ()
  }

  it should "handle transmission errors" in {
    ReliableTransmission.handleTransmissionError("peer1", "Test error")
    // Assuming some mechanism to verify the error was handled
    // This is a placeholder for actual verification logic
    true should be (true)
  }

  it should "restart the reliable transmission service" in {
    val result = Await.result(ReliableTransmission.restart(), 5.seconds)
    result shouldBe ()
  }
}
